### ISSUE TITLE
 cdc: allow sending writes to the previous generation for some time after switching

### ISSUE THREAD

**Issue by kbr-** [scylladb/scylladb#7251]
Currently a node which becomes a write coordinator allows the write to proceed only if its timestamp belongs to the currently operating CDC generation (i.e. the one with the latest timestamp).

The reason for this restriction is preserving colocation. We "guarantee" (or plan to guarantee in the future, when we do consistent topology changes on top of Raft) that the latest generation's token ranges refine the token ring's vnodes, so each base write and its corresponding log write have partition keys whose tokens fall into the same vnode. There is no such guarantee for the previous generation: indeed, that's why we do the generation switch in the first place.

This restriction is not a problem if the client is using server-side generated timestamps, because the generated timestamp will always fall into the latest generation. However, if they're using client-generated timestamps, and clocks are not perfectly synchronized (they are not), there may be short periods of time around the moment of switching generations that client writes will be rejected because they fall into the previous generation. The client gets the following message:
```
    if (it->first > ts) {
        throw exceptions::invalid_request_exception(format(
                "cdc: attempted to get a stream from an earlier generation than the currently used one."
                " With CDC you cannot send writes with timestamps too far into the past, because that would break"
                " consistency properties (write timestamp: {}, current generation started at: {})",
                format_timestamp(ts), format_timestamp(it->first)));
```

This problem is very easy to overcome by the client: simply repeat the write a few times, but using a higher timestamp. Eventually, usually after a few milliseconds, the write will succeed.

Unfortunately, if the client leans on the driver's retry policy instead of doing the retries themselves, the driver will retry the write with the same timestamp. So it's a bit of a nuisance.

---

In the future, we will ensure that the generation is switched before new tokens are introduced into the token ring. We can also ensure that these two events are separated by a few seconds, so that the old generation still has the colocation property during this short time interval. Thus, during these few seconds we could allow writes to be sent to the previous generation without losing any consistency. This will be tricky to implement and assumes bounded clock drifts, but it should be possible.

Today we don't have colocation during node bootstraps anyway; the generation switch is racing with the token ring update. So we could just relax the restriction a bit without worrying too much anyway e.g. by allowing the writes to be sent to the previous generation if the new generation was introduced less than 5 seconds ago, or something like that.

**Comment by kbr-** [comment]
@haaawk @elcallio @nyh @piodul 
perhaps you have some opinions on the issue

**Comment by haaawk** [comment]
I think this should be a property user can pick.
If a user cares very much about better consistency then they may prefer the exception.
Otherwise, they may be fine with writing to the old generation.
I would suggest making the tolerance window (those 5 seconds @kbr- mentioned) a CDC table parameter.
This way it will be flexible enough to serve a wide range of use cases.
Probably the right time unit of such a parameter would be `ms` instead of `s`.

**Comment by slivne** [comment]
I don't think we should write to previous generations - I would prefer we write to the current generation.

The client side reading - using the window should be smart enough to handle reading info from two generations that is not perfectly aligned by time and merge them for a correct view.

**Comment by haaawk** [comment]
@avikivity didn't want clients to merge streams and I agree with him. With a distributed client one worker can read the stream that contains a given partition in the previous generation and totally different worker can read the stream that contains the same partition in the current generation. Those can be totally different machines and merging would be very hard and error-prone. All guarantees we give is that if the client ensures that the previous generation was fully read before it starts reading the next generation, changes will be in order.

I believe that writing to the current generation can lead to a situation where we will end up with the wrong order of records:

1. We have to generations G1 and G2.
2. Generation G1 ended at time T.
3. Generation G2 started at time T.
4. A write W1 with client timestamp equal to T - 500ms arrives
5. But there is already a write W2 with timestamp T - 100ms in generation G1
6. If we write W1 to G2 then we will violate an invariant that reading all records from all streams in G1 gives the client guarantee that it can safely read G2 and all the records will be in the right order.

What's the reason you don't want to write to the old generation @slivne? Clients read records from CDC Log with 10s lag so we can safely add records to the old generation if they are no older than few seconds.

**Comment by kbr-** [comment]
>  I would suggest making the tolerance window (those 5 seconds @kbr- mentioned) a CDC table parameter.

We have to make it bounded if we want to preserve colocation.

My point was that we **don't have to lose colocation** if we wait for some time after we switch the generation but before we introduce new tokens to the token ring. This time interval will be a global property of the cluster.  During this time interval the cluster still uses old token ranges, hence we can write to the old generation with colocation preserved.

> I believe that writing to the current generation can lead to a situation where we will end up with the wrong order of records:

Yes, and preventing this was the whole point of having generations, generation timestamps, and so on.

**Comment by kbr-** [comment]
Here's what we could do when we introduce new tokens to the token ring:
1. create a new CDC generation that is good for both old and new token rings
2. announce it, but allow all nodes to write both to the old and the new generation
3. wait until everyone's clock is past the new generation timestamp
4. wait 5 seconds
5. tell everyone not to write to the previous generation anymore
6. wait until everyone acknowledges 
7. announce the token ring change

So we don't even need clock assumptions. But now we guaranteed that for each coordinator there are at least 5 seconds after it has already entered the new generation during which it is still safe to write to the old generation without losing colocation (because the token ring hasn't changed yet). During these 5 seconds we don't have to reject client writes with timestamps that are a bit off w.r.t the coordinator's clock.

**Comment by haaawk** [comment]
What does it mean to create a new CDC generation that is good for both old and new token rings?

**Comment by haaawk** [comment]
I don't get this idea of writing to both the old and new generation at the same time. It will create duplicated records for a single change, no?

**Comment by kbr-** [comment]
Say new generation G2, has timestamp `T`, so G1 closes at `T`. Say coordinator's clock is at `C >= T`, so the coordinator entered the new generation.

Today if we get a write with timestamp `T - 10ms` we will reject it because it belongs to G1. This never happens with server-side timestamps because the generated timestamp will be `>= C >= T`, but it can easily happen with client timestamps.

I'm proposing a solution where if we get a write with timestamp `T - 10ms`, we allow it and send it to G1, if we can guarantee that G1 is still good for the current token ring. Which we can do for a small interval of time, as shown above. With the example solution above we will allow writes to be sent to G1 as long as `C < T + 5s`, which will still be colocated because we change the token ring only after `C >= T + 5s`.

"Writing to both generations at the same time" means that client writes that belong to G1 go to G1 and client writes that belong to G2 go to G2 (we don't duplicate). Today, client writes that belong to G2 go to G2 and those that belong to G1 get rejected.

> What does it mean to create a new CDC generation that is good for both old and new token rings?

Take the tokens from both token rings and make a stream ID for each token range in the "sum" token ring. This way you get colocation for both token rings.

**Comment by haaawk** [comment]
> I'm proposing a solution where if we get a write with timestamp T - 10ms, we allow it and send it to G1, if we can guarantee that G1 is still good for the current token ring.

What's 'current token ring'? Old? New? Something else?

**Comment by haaawk** [comment]
> Take the tokens from both token rings and make a stream ID for each token range in the "sum" token ring. This way you get colocation for both token rings.

That could mean one vnode has twice the number of streams, no?

**Comment by haaawk** [comment]
> Here's what we could do when we introduce new tokens to the token ring:
> 
> 1. create a new CDC generation that is good for both old and new token rings
> 2. announce it, but allow all nodes to write both to the old and the new generation
> 3. wait until everyone's clock is past the new generation timestamp
> 4. wait 5 seconds
> 5. tell every node not to write to the previous generation anymore
> 6. wait until everyone acknowledges
> 7. announce the token ring change
> 

Why do we need 5, and 6? Can't each node just accept writes to old generation until the new generation timestamp + 5s?

**Comment by slivne** [comment]
Ok - I understand why we prefer to write to the old and new generations - for client simplicity and to provide the grantees - yet my concerns are:
-  if we have a node restart in a middle of a change then everything becomes more complex to validate and test.
- The co-locality assumption is invalid if we write into old and new generations. When we add a new node into the cluster we end up with a new topology - while we guarantee that the active topology and active streams are co-located on the same node and same shard, we can't guarantee that for the old streams - we will end up with CDC writes being done to another node (and all the issues related to that - e.g. writes may fail).
- Looking into the future - I expect we will not support this under raft/tablets (we may not support client side timestamps) - if that is the case - I am not sure why we should attempt to support this now. Client side timestamps are indeed recommended to solve issues - but they are not supported for LWT - if its clear that we will not be able to support this under tablets we need to think of a better solution.

Allowing clients to write in the past also breaks other guarantees - ability to provide correct pre-image as an example.

As such maybe its more correct - todo the same that we do with LWT writes and provide a server timestamp in case the client is requesting CDC to be enabled on a table (we do that in LWT and users accept the why).

**Comment by slivne** [comment]
For CDC GA we need to decide what we will do and can't push the decision

**Comment by haaawk** [comment]
> Ok - I understand why we prefer to write to the old and new generations - for client simplicity and to provide the grantees - yet my concerns are:
> 
> * if we have a node restart in a middle of a change then everything becomes more complex to validate and test.

Could you elaborate on this one @slivne, please?

> * The co-locality assumption is invalid if we write into old and new generations. When we add a new node into the cluster we end up with a new topology - while we guarantee that the active topology and active streams are co-located on the same node and same shard, we can't guarantee that for the old streams - we will end up with CDC writes being done to another node (and all the issues related to that - e.g. writes may fail).

That's why @kbr- wrote in point 1 to generate a generation that's co-located with both old and new token ring. This means it is created with all the tokens from both rings taken into consideration. This can lead to a bit bigger number of streams but should work.

> * Looking into the future - I expect we will not support this under raft/tablets (we may not support client side timestamps) - if that is the case - I am not sure why we should attempt to support this now. Client side timestamps are indeed recommended to solve issues - but they are not supported for LWT - if its clear that we will not be able to support this under tablets we need to think of a better solution.

@kbr- please comment on this one

> 
> Allowing clients to write in the past also breaks other guarantees - ability to provide correct pre-image as an example.

Without Raft the pre-image really gives no guarantees.

> 
> As such maybe its more correct - todo the same that we do with LWT writes and provide a server timestamp in case the client is requesting CDC to be enabled on a table (we do that in LWT and users accept the why).

This is an interesting idea which would make lots of things simpler.

**Comment by elcallio** [comment]
Fwiw, allowing writing to an old generation also has the drawback of data existing in that generation past its expire time, which makes it more difficult to determine when data is exhausted in a particular set of streams. (Obviously can add the window into the select span, but it is yet another place where one has to know magic values to be able to read from cdc log correctly). 

+1 for the idea to forbid client timestamps for cdc enabled writes.

**Comment by kbr-** [comment]
> What's 'current token ring'? Old? New? Something else?
What do you mean by G1 is still good for the current token ring? Colocated?

For generation G and token ring R, define G to be **good** for R if G has at least one stream ID for each vnode in R, and tokens inside those vnodes from R are mapped to these stream IDs so that each log write performed using G and R is colocated with its corresponding base write.

Here's what we do today on  token ring change:
1. token ring is T1
2. generation is G1
3. a node starts bootstrapping and wants to change T1 to T2. In this case T2 is T1 + additional tokens
4. before it changes to T2, the bootstrapping node first creates a generation G2 that is good for T2 and announces it. Because T2 is T1 + additional tokens, G2 is also good for T1.
5. the node then announces T2 without any waiting

> That could mean one vnode has twice the number of streams, no?

Note that I'm not proposing to change how we choose the new generation, we already do it this way.
And it won't have twice the number of streams because the tokens of T1 are a subset of the tokens of T2 when a node is joining. And when a node is leaving we don't change generations.
If we ever do arbitrary token ring changes in the future (when all old tokens are gone) then potentially yes, each vnode in the new ring could have even twice the number of necessary streams.

> Why do we need 5, and 6? Can't each node just accept writes to old generation until the new generation timestamp + 5s?

I just proposed an example solution, your suggestion would probably work as well (however your solution assumes synchronized clocks). And I'm sure there are even more different solutions.

**Comment by kbr-** [comment]
> The co-locality assumption is invalid if we write into old and new generations. When we add a new node into the cluster we end up with a new topology - while we guarantee that the active topology and active streams are co-located on the same node and same shard, we can't guarantee that for the old streams - we will end up with CDC writes being done to another node (and all the issues related to that - e.g. writes may fail).

Consider the following problem:
1. for token ring T1, generation G1 provides colocation
2. for token ring T2, generation G2 provides colocation
3. for T1, G2 doesn't provide colocation
4. for T2, G1 doesn't provide colocation

We currently use T1 and want to switch to T2. We have two options:
1. first switch T1 to T2, then switch G1 to G2.
But then there's a period of time when T2 operates together with G1, so no colocation
2. first switch G1 to G2, then switch T1 to T2
But there's a period of time when T1 operats with G2, so no colocation

and it's not possible to switch "at the same time" because reasons (the entire thing is distributed across multiple nodes and so on).

We **already solve that problem today** by actually ensuring that G2 is colocated with both T1 and T2. So we can first switch G1 to G2, and then switch T1 to T2, and we will have colocation at all times. Now I propose to do the following:
1. switch G1 to G2
2. wait 5 seconds; during these 5 seconds nodes are still allowed to write to G1 if the write belongs to G1 (because of its client timestamp)
3. then switch T1 to T2

It's safe to write to G1 during (2) because we still use ring T1.
Of course after the 5 seconds we're no longer guaranteed to use T1, so we would reject writes that belong to G1 (and it's OK because it's very unlikely that such writes will occur unless the client is crazy).

> Looking into the future - I expect we will not support this under raft/tablets (we may not support client side timestamps) - if that is the case - I am not sure why we should attempt to support this now. Client side timestamps are indeed recommended to solve issues - but they are not supported for LWT - if its clear that we will not be able to support this under tablets we need to think of a better solution.

For tablets everything's going to have to be reworked since they will use a different partitioning scheme (AFAIK), not based on the cluster-global token ring, and each tablet will have its own partitioning, so we may have to use a separate CDC "generation" per each table. So we'll most probably be able to make different decisions then.

Client timestamps aren't really a problem -- as I said in the original post, we already reject the write if the timestamp is invonvenient for us, and we can just leave it as it is, the client has to retry the write in that case.

> Allowing clients to write in the past also breaks other guarantees

This is an orthogonal issue; writes are never guaranteed to have increasing timestamps, out-of-order writes happen even within a single CDC generation.

> This is an interesting idea which would make lots of things simpler.

Client timestamps are easy to handle by simply saying "wrong timestamp, try again with a better timestamp" which we do today. We can just leave the status quo and close the issue, we don't have to prevent client timestamps completely.

> Fwiw, allowing writing to an old generation also has the drawback of data existing in that generation past its expire time, which makes it more difficult to determine when data is exhausted in a particular set of streams. (Obviously can add the window into the select span, but it is yet another place where one has to know magic values to be able to read from cdc log correctly).

Whether the generation is expired is determined on each node locally using that node's clocks, and different nodes decide at different times whether the generation is switched. To correctly handle generation changes today, you have to use the window approach anyway. It's documented here: https://docs.scylladb.com/using-scylla/cdc/cdc-querying-streams/
```
Keep in mind that time is relative: every node has its own clock. Therefore you should make sure that the old generation stops operating from the point of view of every node in the cluster before you query it one last time and start querying the new generation.
```

preventing client timestamps won't help that.

**Comment by kbr-** [comment]
We concluded on a call that it's safest if the current restrictions stay. The client can either use server-timestamps or, if their request gets rejected due to belonging to an earlier generation, simply repeat it with a higher timestamp (note that with this error the client knows that  the write was a definite failure).

**Comment by kbr-scylla** [comment]
> We concluded on a call that it's safest if the current restrictions stay.

And today, our customers are hitting "attempted to get a stream from an earlier generation than the currently used one" in production and our engineers are getting paged at night.

https://github.com/scylladb/scylla-enterprise/issues/3776

I think we should do something at least to minimize it from occurring. For example, allow a write to the previous generation if the write's timestamp is not too far into the past (e.g. `T > now - 5s`).

cc @mykaul @avelanarius

**Comment by kbr-scylla** [comment]
> The client can either use server-timestamps or, if their request gets rejected due to belonging to an earlier generation, simply repeat it with a higher timestamp (note that with this error the client knows that the write was a definite failure).

The above reasoning is correct... with one exception -- LWT.

The problem is that LWT **makes timestamps permanent**. Once Paxos commits an entry with a given timestamp, Scylla will keep trying to apply that entry until it succeeds, with the same timestamp (using a different timestamp would potentially introduce inconsistency -- we **must** use the originally committed timestamp).

But then applying the entry involves doing a CDC log table write, if CDC is enabled on the table. And this part may fail, if the timestamp belongs to a previous generation.

Whenever we retry an LWT query for a given (partition, table) pair, the Paxos logic will first start by trying to reapply the existing entry... which will fail again due to the old timestamp. So we're stuck.

So, even though the problem is indeed transient and non-serious outside LWT, when we combine LWT with CDC, we get a serious bug.

**Comment by gleb-cloudius** [comment]
> > We concluded on a call that it's safest if the current restrictions stay.
> 
> And today, our customers are hitting "attempted to get a stream from an earlier generation than the currently used one" in production and our engineers are getting paged at night.
> 
> [scylladb/scylla-enterprise#3776](https://github.com/scylladb/scylla-enterprise/issues/3776)
> 
> I think we should do something at least to minimize it from occurring. For example, allow a write to the previous generation if the write's timestamp is not too far into the past (e.g. `T > now - 5s`).

But with LWT, theoretically, it can be much older. BTW why hints do not have the same problem?

**Comment by kbr-scylla** [comment]
CDC write is performed at the same moment as the base table write is coordinated.
Sending hints doesn't trigger cdc writes. Hints correspond to writes that were coordinated in the past.

**Comment by nyh** [comment]
@kbr-scylla I'm afraid to ask because I didn't spend enough time on this, but could this issue have the same root cause as https://github.com/scylladb/scylladb/issues/15260? 
In that other issue Alternator Streams tests (which use CDC, of course), sporadically fail with "cdc: attempted to get a stream from an earlier generation than the currently used one." and we weren't able to figure out why.

**Comment by kbr-scylla** [comment]
@nyh In https://github.com/scylladb/scylladb/issues/15260 as far as I understand it should not happen because:
- the test is not creating new generations while it is running (there are no node bootstraps, it is a single test), there is only the one generation, the first one
- and the test supposedly waits in some way for that first generation's timestamp (?)

The question is how the wait is implemented. To avoid the error, the wait should ensure that any writes started by the test will have their timestamps greater than the generation's timestamp.

Now unfortunately I think that fixing *this* issue will not help in the case of https://github.com/scylladb/scylladb/issues/15260. The proposal in this issue is to allow sending writes to "previous" generations if they are not too far off current time. But in the case of https://github.com/scylladb/scylladb/issues/15260, there is no previous generation -- the problem is already with the first generation, and apparently the test is trying to do a write with a timestamp which is smaller than the first generation's timestamp. So if anything, the error will just be converted to another one. You cannot send a write to non-existing generation.

> @kbr-scylla - do we have a high level at least design for a solution for this?

Basically what I wrote [here](https://github.com/scylladb/scylladb/issues/7251#issuecomment-1882787210)

> I think we should do something at least to minimize it from occurring. For example, allow a write to the previous generation if the write's timestamp is not too far into the past (e.g. T > now - 5s).

This is a probabilistic fix which should get rid of the problem in practice. Other than that, I have absolutely no idea for a full 100% theoretical fix (i.e. if the clocks go more than 5 seconds, we'll still get the error) -- the problem is deeply ingrained into the design of CDC.

**Comment by kbr-scylla** [comment]
@nyh ok I messed some facts up, the test actually **is** bootstrapping a node in the middle.

So yes, the fix should help.

**Comment by tzach** [comment]
@kbr-scylla any update on this?

**Comment by kbr-scylla** [comment]
@patjed41 is working on it, he might provide an update

**Comment by patjed41** [comment]
It's work in progress.

One update I can give is that the solution I'm going for is a bit more complicated than the one proposed [here](https://github.com/scylladb/scylladb/issues/7251#issuecomment-1882787210). The difference is "previous generation" -> "previous generation**s**".

The proposed solution fixes writes only to the previous generation. Writes to generations older than the previous generation would still fail if `now - 5s < T < now`. Currently, it shouldn't be a problem, because differences between CDC generation timestamps are usually big, since they are created when we add nodes. However, in the Raft-based topology, we can add multiple nodes concurrently. After moving to tablets, streaming will become instant and the topology coordinator will be able to add multiple nodes very quickly. Then, consecutive CDC generations with almost identical timestamps will appear, and we might encounter the error from https://github.com/scylladb/scylla-enterprise/issues/3776, but, for example, for the generation before the previous generation.

**Comment by mykaul** [comment]
It'd be great to have a short design document to the solution, instead of following endless comments in this GH issue (once agreed upon)
